### DISCUSSION TITLE
How to switch shaders at runtime?

### DISCUSSION THREAD

**Discussion by yangcshen** [vsg-dev/VulkanSceneGraph#803] [General]
Hi 

I am using vsgXchange built with assimp to read 3D model files such as stl and obj. I would like to have different visualizations such as displaying all the edges in a mesh and display normal vectors. My understanding is that I need to use a different shader and perhaps change some configurations in graphics pipeline. How do I do that dynamically at runtime? Or I am asking the wrong question. I was looking into vsgExample. Which one is the best example? 

Thanks,
Cam

**Comment by robertosfield** [comment]
Have a look at vsgstateswitch:

https://github.com/vsg-dev/vsgExamples/blob/master/examples/state/vsgstateswitch/vsgstateswitch.cpp

This example locates the GrpahicsPipeline in the scene graph and then uses a StateSwitch to enable each View to have it's own state, with the left view using the original state, and the right view using a wireframe version,

**Comment by yangcshen** [comment]
Circling back here as I have made some progress. I have some questions regarding the use of vsg::Operation + vsg::OperationThreads.
1. My understanding is that if I need to change something at runtime (meaning in a new rendering loop), I need to use this pattern. sgdynamicload and vsgdynamicviews are good examples. Is this correct? If I want to keep everything on my main thread, what should I do? addNode -> compile -> update viewer?
2. I see that in the examples it is always two structs (operation + merge). I do not fully understand it. Can I combine them into one? What is the reason having this pattern?

**Reply by robertosfield** [reply]
Could you create a new thread for this as it's a new topic, this will make it easier for others to find and contribute to the discussion.

**Reply by yangcshen** [reply]
#805
